### PR TITLE
fix(2048): incorrect action mask

### LINKED ISSUE #127: bug(2048): action_mask calculated before adding random cell
### Description

The action mask returned by Game2048 is calculated before adding the random cell, which leads to incorrect action masks.
```
env = Game2048()
state = State(
    board=jnp.array(
        [[0, 1, 2, 3], [3, 1, 2, 3], [1, 2, 3, 4], [4, 3, 2, 1]], dtype=jnp.int32
    ),
    step_count=jnp.array(0, dtype=jnp.int32),
    action_mask=jnp.array([True, False, True, True], dtype=jnp.bool_),
    score=jnp.array(0.0, dtype=jnp.float32),
    key=jnp.array([0, 0], dtype=jnp.uint32),
)
next_state, _ = env.step(state, 3)
print(next_state.action_mask)
```
This code outputs `[True, True, False, False]`, but it should output `[False, False, False, False]`. In a related error, a random cell will be added even if the previous action failed.
```
env = Game2048()
state = State(
    board=jnp.array(
        [[1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], dtype=jnp.int32
    ),
    step_count=jnp.array(0, dtype=jnp.int32),
    action_mask=jnp.array([False, True, True, True], dtype=jnp.bool_),
    score=jnp.array(0.0, dtype=jnp.float32),
    key=jnp.array([0, 0], dtype=jnp.uint32),
)
next_state, _ = env.step(state, 0)
print(next_state.board)
```
This code outputs `[[1, 1, 1, 1], [0, 0, 1, 0], [0, 0, 0, 0], [0, 0, 0, 0]]`, but it should output `[[1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]`.

### What Jumanji version are you using?

jumanji 0.2.2

### Which accelerator(s) are you using?

CPU/GPU

### Additional System Info

3.11

### Additional Context

_No response_

### (Optional) Suggestion

The action_mask and done variables should be calculated after adding the random cell and the condition to add the random cell should be changed to `state.action_mask[action]`.

### PR DESCRIPTION
**Pull request by aar65537**
closes #127

- Fixes issue where action_mask is calculated before adding random tile
- Fixes inconsistency with original game where performing invalid action still adds a random tile
- Includes unit tests verifying fixes
- Increments Game2048 version in registry because changes environment behavior

### CODE REVIEW

File: jumanji/environments/logic/game_2048/env.py [hunk 1]
```diff
@@ -187,25 +187,22 @@ def step(
...
         # Generate new key.
         random_cell_key, new_state_key = jax.random.split(state.key)
 
         # Update the state of the board by adding a new random cell.
         updated_board = jax.lax.cond(
-            done,
-            lambda board, key: board,
+            state.action_mask[action],
             self._add_random_cell,
+            lambda board, key: board,
```
Comment (OmaymaMahjoub) [inline jumanji/environments/logic/game_2048/env.py:197]:
```suggestion
```

File: jumanji/environments/logic/game_2048/env_test.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ def test_game_2048__step_jit(game_2048: Game2048) -> None:
     """Confirm that the step is only compiled once when jitted."""
```
Comment (OmaymaMahjoub) [inline jumanji/environments/logic/game_2048/env_test.py:61]: Nice efforts in adding additional tests :raised_hands: , however, I think no need to add these tests since we checked the action masking `test_game_2048__get_action_mask`. I don't if @clement-bonnet  do you think we need to fix these tests and add them (maybe we can keep `test_game_2048__step_action_mask `) ?

File: jumanji/environments/logic/game_2048/env_test.py [hunk 3]
```diff
@@ -75,12 +75,47 @@ def test_game_2048__step_jit(game_2048: Game2048) -> None:
...
+        board=jnp.array([[1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]),
+        step_count=jnp.array(0),
+        action_mask=jnp.array([False, True, True, True]),
+        score=jnp.array(0),
+        key=jax.random.PRNGKey(0),
+    )
+    action = jnp.array(0)
+    step_fn = jax.jit(game_2048.step)
+    new_state, next_timestep = step_fn(state, action)
+    assert jnp.array_equal(state.board, new_state.board)
```
Comment (OmaymaMahjoub) [inline jumanji/environments/logic/game_2048/env_test.py:97]: They can't be equal since we have the random cell that appear each cell

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Thanks @aar65537  for your contribution in fixing the issue that you noticed :pray:  I made here some minor points to be fixed to make your solution more efficient :slightly_smiling_face:

**Review by clement-bonnet — APPROVED** [review]
Nice catch! Thank you for the fix!

File: README.md [hunk 4]
```diff
@@ -78,7 +78,7 @@ problems.
 
 | Environment                              | Category | Registered Version(s)                                | Source                                                                                           | Description                                                            |
 |------------------------------------------|----------|------------------------------------------------------|--------------------------------------------------------------------------------------------------|------------------------------------------------------------------------|
-| 🔢 Game2048                              | Logic  | `Game2048-v0`                                        | [code](https://github.com/instadeepai/jumanji/tree/main/jumanji/environments/logic/game_2048/)   | [doc](https://instadeepai.github.io/jumanji/environments/game_2048/)   |
+| 🔢 Game2048                              | Logic  | `Game2048-v1`                                        | [code](https://github.com/instadeepai/jumanji/tree/main/jumanji/environments/logic/game_2048/)   | [doc](https://instadeepai.github.io/jumanji/environments/game_2048/)   |
```
[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply README.md:81]: I wasn't aware about this fact :sweat_smile: I will close this issue
